### PR TITLE
uudecoding functions

### PR DESCRIPTION
**Pull request by csb0730**
Hi @r10s,

changes easier than thought. FILE related functions really has been unnecessary :-)

Only _dc_uudecode.c_ is updated.

### CODE REVIEW

**Comment by csb0730** [comment]
@r10s - full rework of `dc_uudecode.c` done and tested locally. May You make me a new apk again?

**Comment by csb0730** [comment]
latest apk self done :)

**Comment by ralphtheninja** [comment]
@csb0730 Is this ready for review or are you just playing around?

**Comment by csb0730** [comment]
Ready for review.
I completed individual tests with expected results.
All expectations from my side are fulfilled :)

**Comment by ralphtheninja** [comment]
Please rebase this on latest master and clean up unnecessary comments and debug code. There's also a lot of style differences compared to the rest of the code base.

**Comment by csb0730** [comment]
There is a patch-1 to the PR now.
Somehow it was not possible to create a new fork from dc core master so I did it from the existing fork.
In patch-1 I uploaded the updated `dc_uudecode.c` as requested.

**Comment by csb0730** [comment]
Do I need to delete the existing csb7030 fork before creating a new one from original dc core repo?
I did this as an emergency action last time but I intended to create a new fork and keep the old one too.

**Comment by csb0730** [comment]
Ok, achieved now a merge to include patch-1 into my own master (rebase and merge function). It seemed to work :)

**Comment by csb0730** [comment]
Is the code ok now or are further changes/formatting requested?
(only `dc_uudecode.c` is changed from base branch).

File: src/dc_uudecode.c [hunk 1]
```diff
@@ -32,31 +86,564 @@
...
+	}
+	
+	// here: no uuencoded text part identified
+	return NULL;
+	
+} //dc_find_uuencoded_part()
+
+
+char* dc_handle_uuencoded_part (const char*	msgtxt,
+								 char*			uu_msg_start_pos,
```
Comment (r10s) [inline src/dc_uudecode.c]: nitpicking: tabs can only be used for indentation, for alignment, spaces are needed.
(maybe we should switch to spaces doe indentation as well, but this is a different discussion :)

File: src/dc_uudecode.c [hunk 2]
```diff
@@ -32,31 +86,564 @@
...
+	
+} //dc_print_hex()
+
+
+
+int dc_getline (char** ret_lineptr, char* source, char** ret_nextcharptr){
+	/**
+	 * 	Extracts first line from a given string (source)
+	 * 
+	 * 	Instead of strtok_r it doesn't manipulate the input string
```
Comment (r10s) [inline src/dc_uudecode.c]: so, input and outout can be const?

File: src/dc_uudecode.c [hunk 3]
```diff
@@ -32,31 +86,564 @@
...
+
+int dc_getline (char** ret_lineptr, char* source, char** ret_nextcharptr){
+	/**
+	 * 	Extracts first line from a given string (source)
+	 * 
+	 * 	Instead of strtok_r it doesn't manipulate the input string 
+	 *
+	 *  Parameters:
+	 * 
+	 *  @ret_lineptr   - returned line incl. line end char
```
Comment (r10s) [inline src/dc_uudecode.c]: i think doxygen won't accept this format, it should read `@param ret_lineprt Explain line pointer`

**Comment by csb0730** [comment]
Hi @r10s 
Many hints!
I'll have a look into all this. But because the code is working fine maybe its a quicker approach if You prepare an updated file with requested changes so that I can see what's needed?

**Comment by csb0730** [comment]
A little too quick button pressed ;-)

**Comment by r10s** [comment]
@csb0730 i would prefer to stay here on github. also because others can handle/review this as well.

File: src/dc_uudecode.c [hunk 4]
```diff
@@ -20,10 +20,64 @@
...
+
+
 #include "dc_context.h"
+#include "dc_tools.h"
+
 #include "dc_uudecode.h"
 
 
+/**
+ *  Locally needed function declarations
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: Does it mean that these declarations all should be moved to "dc_uudecode.h" and not doing declaration in .c file?
Or only declare all local functions with static keyword where they are now?

File: src/dc_uudecode.c [hunk 5]
```diff
@@ -32,31 +86,564 @@
...
+ *     The caller will call mruudecode_do() again with this remaining text then.
  *     This way, multiple uuencoded parts can be stripped from a text.
  *     If no uuencoded parts are found or on errors, NULL is returned.
  */
 char* dc_uudecode_do(const char* text, char** ret_binary, size_t* ret_binary_bytes, char** ret_filename)
 {
 	// CAVE: This function may be called in a loop until it returns NULL, so make sure not to create an invinitive look.
 
-	if (text == NULL || ret_binary == NULL || ret_binary_bytes == NULL || ret_filename == NULL) {
+	if( text == NULL || ret_binary == NULL || ret_binary_bytes == NULL || ret_filename == NULL ) {
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: That's no problem, readability is important, agree!
Didn't pay attention to that while testing ;-)

File: src/dc_uudecode.c [hunk 6]
```diff
@@ -32,31 +86,564 @@
...
+
+	// find again in new buffer
+	uustartpos = dc_find_uuencoded_part(txt);
+	if (uustartpos){
+		// then handle uuencoded part
+		ret_msg_text = dc_handle_uuencoded_part (txt, uustartpos, ret_binary, ret_binary_bytes, ret_filename);
+	}
+	free(txt);
+	
+	return ret_msg_text;
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: Means "return ..." only in cleanup section? Good approach. Will change that.

File: src/dc_uudecode.c [hunk 7]
```diff
@@ -32,31 +86,564 @@
...
+	if (!uustartpos){
+		// no or no further uuencoded parts found
+		goto cleanup;
+	}
+
+	/**
+	 * optimization possible: identify first call of this function to unify buffer only once
+	 */
+	// prepare buffer and unify buffer
+	char* txt = strdup(text);
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: Makes sense -> change

File: src/dc_uudecode.c [hunk 8]
```diff
@@ -20,10 +20,64 @@
...
+/* decode uuencoded part and provide it, used in dc_handle_uuencoded_part() */
+int   dc_uudecode(char** ret_binary, size_t uudecoding_buffer_len, const char* uu_body_start);
+
+
+
+/**
+ *  Variables and preprocessor definitions
+ */
+
+#define LF   "\n"   //10
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: Used only once: making it local

File: src/dc_uudecode.c [hunk 9]
```diff
@@ -32,31 +86,564 @@
...
+		
+		linelen = lineend - start; // len of line
+
+		*ret_nextcharptr	= lineend;
+		
+		*ret_lineptr		= strndup(start, linelen);
+	}
+	
+	return linelen;
+} //dc_getline()
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: Going to standard, this is better :)

File: src/dc_uudecode.c (hunk 1 above)
Comment (csb0730) [reply src/dc_uudecode.c]: -> Doing update

File: src/dc_uudecode.c (hunk 2 above)
Comment (csb0730) [reply src/dc_uudecode.c]: To check ...

File: src/dc_uudecode.c (hunk 3 above)
Comment (csb0730) [reply src/dc_uudecode.c]: Ok, now I understand how this is used. Will update the function descriptions.

**Comment by csb0730** [comment]
In general all points make sense. I will update the code and will reload it here.

By the way, the first PR I deleted because I wanted to pull the newest code here to add the new dc_uudecode.c. Creating a second PR while another one is active seems to be impossible.
Because of the last comments of @ralphtheninja I tried the same but fortunately that didn't succeeded here so the PR is still active and the comments are not deleted.
Up to now I still do not know how how doing a pull here from original source repo! Mabe someone can give me some hints how to achieve that.

File: src/dc_uudecode.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (csb0730) [reply src/dc_uudecode.c]: See comment why 1st PR has been deleted/archived below.

[the remark this answers is not in the document]
Comment (r10s) [reply src/dc_uudecode.c]: i would not move the functions to an additional header as they are only used locally.
i would just move them "up" and call them further "down".

> Or only declare all local functions with static keyword where they are now?

basically, yes.

**Comment by r10s** [comment]
@csb0730 i am no git expert, but what you're doing in the pr is basically correct, you have your fork on `csb0730:master` and do a pr from there to `deltachat:master`.

what i simply suggest is, that this pr is  not closed on changes and a new one is created.
i think it should be possible to add commits to your fork and push them, i think they will appear here in this pr then.

with a `git rebase` and a `git push --force` you can cleanup things, but iirc, you have already done this some time :)

File: src/dc_uudecode.c [hunk 10]
```diff
@@ -20,10 +20,64 @@
...
+/**
+ *  Variables and preprocessor definitions
+ */
+
+#define LF   "\n"   //10
+
+
+/**
+ *  Function defintions
+ */
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `defintions` -> `definitions`

File: src/dc_uudecode.c [hunk 11]
```diff
@@ -32,31 +86,564 @@
...
+	if( text == NULL || ret_binary == NULL || ret_binary_bytes == NULL || ret_filename == NULL ) {
 		goto cleanup; // bad parameters
 	}
 
+	char* uustartpos    = NULL;
+	char* ret_msg_text  = NULL; // NULL = no uuencoded parts are found (standard case)
+	
+	// first make a quick check if a part exists
+	uustartpos = dc_find_uuencoded_part(text);
+	if (!uustartpos){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `) {`

File: src/dc_uudecode.c [hunk 12]
```diff
@@ -32,31 +86,564 @@
...
+	/**
+	 * optimization possible: identify first call of this function to unify buffer only once
+	 */
+	// prepare buffer and unify buffer
+	char* txt = strdup(text);
+	dc_unify_lineends(txt);
+
+	// find again in new buffer
+	uustartpos = dc_find_uuencoded_part(txt);
+	if (uustartpos){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `) {`

File: src/dc_uudecode.c [hunk 13]
```diff
@@ -32,31 +86,564 @@
...
+		// then handle uuencoded part
+		ret_msg_text = dc_handle_uuencoded_part (txt, uustartpos, ret_binary, ret_binary_bytes, ret_filename);
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: Remove space between function and `(`

File: src/dc_uudecode.c [hunk 14]
```diff
@@ -32,31 +86,564 @@
...
+	 *  realloc ret_binay to necessary size and
+	 *  copy error text to ret_binary
+	 * 
+	 *  @return - len of error txt
+	 */
+	 
+	int 	ret_bytes = 0;
+	char 	msg[strlen(format)+33]; // for 3 parameters
+	
+	switch(nargs){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: Spacing.

File: src/dc_uudecode.c [hunk 15]
```diff
@@ -32,31 +86,564 @@
...
+		case 2: sprintf(msg, format, n1, n2); break;
+		case 3: sprintf(msg, format, n1, n2, n3); break;
+		default:
+			return ret_bytes;
+	}
+
+	int      l = strlen(msg);
+	*ret_binary = (char*) realloc(*ret_binary, l+1);
+	
+	if (*ret_binary){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `) {`

File: src/dc_uudecode.c [hunk 16]
```diff
@@ -32,31 +86,564 @@
...
+		// write error msg to output file if possible!
+		strncpy(*ret_binary, msg, l);
+		ret_bytes = l * (-1);
+	}
+	else{
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `else {`

File: src/dc_uudecode.c [hunk 17]
```diff
@@ -32,31 +86,564 @@
...
+		ret_bytes = 0;
+	}
+	
+	return ret_bytes;
+}
+
+
+char* dc_print_hex (char* s){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `{` should start at the next line.

File: src/dc_uudecode.c [hunk 18]
```diff
@@ -32,31 +86,564 @@
...
+	 * 
+	 * @Return
+	 *      hex representation of s (needs to be free'd)
+	 * 
+	 */
+	int n = strlen(s);
+	int i;
+	char* hex = (char*)calloc((2*n+1), sizeof(char));
+	
+	for(i = 0; i<n; i++){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: Spacing.

File: src/dc_uudecode.c [hunk 19]
```diff
@@ -32,31 +86,564 @@
...
+		sprintf(hex+i*2, "%02X", s[i]);
+	}
+	
+	return hex;
+	
+} //dc_print_hex()
+
+
+
+int dc_getline (char** ret_lineptr, char* source, char** ret_nextcharptr){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `int dc_getline(`

File: src/dc_uudecode.c [hunk 20]
```diff
@@ -32,31 +86,564 @@
...
+		
+		*ret_lineptr		= strndup(start, linelen);
+	}
+	
+	return linelen;
+} //dc_getline()
+
+
+
+int dc_uu_check_line (int n, int line_len){
```
Comment (ralphtheninja) [inline src/dc_uudecode.c]: `int dc_uu_check_line(int n, int line_len)` and `{` on next line

File: src/dc_uudecode.c (hunk 3 above)
Comment (ralphtheninja) [reply src/dc_uudecode.c]: Also, move outside of function.

**Comment by ralphtheninja** [comment]
@csb0730 In general, just commit and push your changes to this branch. Once you have fixed lines that were commented on, they will be hidden by github and then we can move on with reviewing. Also, it's not needed to change the title of this issue for how many times you have changed something. The commits will show this.

**Comment by r10s** [comment]
had a rough scan over the code, there are still issues, eg. fixed buffer sizes, just to mention one.

as already mentioned in the [corresponding issue](#90), i am not convinced that support of uuencode is really needed at all. most common other MUAs just seem to ignore them and i've not seen a MUA that generates uuencoded attachments out of the box.

but there are more concerns in adding uudecoding stuff to delta. 

- there is the security aspect. having code that is not executed most of the time but can be targeted from outside by a simple mail that is not even visible to the user, are, after all experience, the perfect fertile soil for exploits.

- uuencode is complex and hard to distinguish from normal text messages: doing it "the right way" would result in a lot of additional effort and research - i'm not sure that we should bind our few resources here.

- it may harm the whole mail ecosystem: there is a better standard since decades. supporting legacy stuff may lower the need of other apps to go forward to better and more documented approaches.

i am sorry that this pr and this corresponding issue was hanging around so long, and i do not like to be the guy that has to decide to close it. however, i think it is _for now_ the correct decision and i think we all have learnt many things :)

of course, the discussion does not have to end here, it can be restarted eg. in the support forum.

@csb0730 thanks a lot for the work and the insights, looking forward to see more pr that can be merged then :)

**Comment by csb0730** [comment]
Hi @r10s thanks a lot for the comment. I want to pick up the proposal for doing further discussion at support forum. Just today I achieve to get access to that :-)

I'm absolutely with you: If there is really no need any more to support uuencode then we should drop it. 

In the moment I'm making some trials and examinations with Outlook 2016 (at work), Thunderbird and my colleague where I've got all these uuencoded attachments from!

So I would appreciate to let the official code in the moment as it is in uudecode.c with the dummy function as long as the discussion is finished. So I can use it further on for testing.

To bring the discussion to a reasonable end I intend to open the discussion at support forum (now I have access :-) ) and document all aspects there so that all contributors can share it.

I think DC is a great project and my vision is to make it really capable to handle all common mail standards so that every user never get in trouble with "sometimes" complicated mail standards. Not less and not more :-)
Handling old fashioned uuencoded attachments could be one part "uncomplicated" user feeling with DC :-)

**Comment by r10s** [comment]
> But, that's true, in principle it's really a pure text part without mail header controlled mechanism!

that's what i meant.

**Comment by csb0730** [comment]
Summary will follow in support forum soon.

**Comment by csb0730** [comment]
Because I cannot upload a pdf at forum I do it here:
[uuencode_conclusion.pdf](https://github.com/deltachat/deltachat-core/files/2540663/uuencode_conclusion.pdf)

Hi @r10s my conclusion is the same than your's: We'll not need it any more. If someone really needs it in the future, I'll keep it in backup ;-)

**Pull request closed without merging**